packages/stream-metadata/src/routes/userSpaces.ts

```diff
@@ -0,0 +1,113 @@
+import { FastifyReply, FastifyRequest } from 'fastify'
+import { StreamPrefix, StreamStateView, makeStreamId, isSpaceStreamId } from '@towns-protocol/sdk'
+import { MembershipOp } from '@towns-protocol/proto'
+import { z } from 'zod'
+import { SpaceAddressFromSpaceId, type ISpaceOwnerBase } from '@towns-protocol/web3'
+
+import { getStream } from '../riverStreamRpcClient'
+import { isValidEthereumAddress } from '../validators'
+import { config } from '../environment'
+import { spaceDapp } from '../contract-utils'
+
+const paramsSchema = z.object({
+	userId: z.string().min(1, 'userId parameter is required').refine(isValidEthereumAddress, {
+		message: 'Invalid userId',
+	}),
+})
+
+const CACHE_CONTROL = {
+	200: 'public, max-age=30, s-maxage=3600, stale-while-revalidate=3600',
+	404: 'public, max-age=5, s-maxage=3600',
+}
+
+export async function fetchUserSpaces(request: FastifyRequest, reply: FastifyReply) {
+	const logger = request.log.child({ name: fetchUserSpaces.name })
+	const parseResult = paramsSchema.safeParse(request.params)
+
+	if (!parseResult.success) {
+		const errorMessage = parseResult.error.errors[0]?.message || 'Invalid parameters'
+		logger.info(errorMessage)
+		return reply.code(400).send({ error: 'Bad Request', message: errorMessage })
+	}
+
+	const { userId } = parseResult.data
+
+	logger.info({ userId }, 'Fetching user spaces')
+
+	let stream: StreamStateView | undefined
+	try {
+		const userStreamId = makeStreamId(StreamPrefix.User, userId)
+		stream = await getStream(logger, userStreamId)
+	} catch (error) {
+		logger.error(
+			{
+				err: error,
+				userId,
+			},
+			'Failed to get stream',
+		)
+		return reply.code(500).send('Failed to get stream')
+	}
+
+	if (!stream) {
+		return reply.code(404).header('Cache-Control', CACHE_CONTROL[404]).send('Stream not found')
+	}
+
+	const spaceIds = getUserSpaces(stream)
+	if (!spaceIds) {
+		logger.info({ userId, streamId: stream.streamId }, 'user spaces not found')
+		return reply
+			.code(404)
+			.header('Cache-Control', CACHE_CONTROL[404])
+			.send('user spaces not found')
+	}
+
+	const spaceInfoMap = new Map<string, ISpaceOwnerBase.SpaceStructOutput>()
+	await Promise.all(
+		spaceIds.map(async (spaceId) => {
+			const spaceAddress = SpaceAddressFromSpaceId(spaceId)
+			const spaceInfo = await spaceDapp.spaceOwner.getSpaceInfo(spaceAddress)
+			spaceInfoMap.set(spaceAddress, spaceInfo)
+		}),
+	)
```

Comment:
Wonder if we should add a batch limit here. Probably wont be a issue right now, but it can trigger a lot of requests.
(iirc, the spaceDapp isnt using multicall, so each one is triggering a single request to the rpc?)

Reply:
Also - multi call?

Reply:
seems like this could be problematic, we'll get rate limited by alchemy

Reply:
Yeah that’s what I’m doing with images. Just retruning the URL here so in the client we can pass to a img html component

Reply:
Multicall https://viem.sh/docs/contract/multicall.html